README.md
orxporter
=========

Emoji exporter used for Mutant Standard emoji set

Introduction
------------

Orxporter was created as a rewrite of scripts used to automate some of the
workflow for Mutant Standard. Its purpose remains to be the primary build tool
for Mutant Standard packages, and so its features remain largely constrained
to a single use case.

That said, there has been interest in this tool, as it allows custom builds
or modifications of Mutant Standard packages, and is generic enough to work
for similar projects.

Disclaimer
----------

Orxporter is an open-sourced insider tool, and was never meant to be used at
large. There has been no significant effort to design, test, or follow any
good practice. Bugs at edge cases are known about and ignored. As long as
Mutant Standard is built properly according to Dzuk's wishes, it's unlikely
any features will be added or bugs fixed.

Features
--------

* Declarative language for defining semantics of emoji set
* Color mapping (recoloring)
* Unicode metadata
* SVG and arbitrary size PNG exports
* Supports renderers: inkscape, rendersvg and imagemagick
* Output options including emoji filtering, customisable export directory
  structure and filenaming
* JSON output of emoji set metadata
* Embedding licensing metadata
* Multithreading

Prerequisites
-------------

* Python 3.6+
* Inkscape (optional; for rasterisation)
* ImageMagick (optional; for rasterisation)
* [rendersvg](https://github.com/RazrFalcon/resvg/tree/master/tools/rendersvg)
  (optional; for rasterisation)
* exiftool (optional; for embedding licensing metadata)

Usage
-----

By default, orxporter assumes the manifest is in **./manifest**, the input
directory is the **in** subdirectory of wherever the manifest is, and the
output directory is **out**. Default action is to export the entire emoji set
as shortcode-named SVG files.

Here are the command line options:

* **-h** -- prints the help message
* **-m PATH** -- specifies manifest file path (default: **manifest**)
* **-i PATH** -- specifies input image directory path; note that as build
  files are meant to be self-contained, this is relative to the manifest
  filepath (default: **in**)
* **-o PATH** -- specifies output directory path; this is relative to the
  current directory (default: **out**)
* **-f PATH_EXPR** -- specifies output filenaming for emoji; the following
  formatting codes are supported: **%c** for colormap, **%d** for source
  image's directory within input directory, **%f** for export format used,
  **%s** for shortcode, **%u** for unicode sequence, **%(property_name)** for
  emoji's property; (default: **%f/%s**)
* **-F FORMAT[,FORMAT...]** -- specifies output formats; the following
  formats are supported: **svg**, **png-SIZE** (fe. **png-64**); (default:
  **svg**)
* **-e FILTER** -- only emoji matching the filter condition will be exported;
  multiple filters can be used to narrow selection down but each requires a
  separate **-e** option; filters are specified as **property=val1[,val2...]**
  to match emoji with the property having the value (or one of listed values),
  __property=*__ to match emoji with the property being defined (regardless of
  value), or **property=!** to match emoji with the property being undefined
* **-j FILE** -- export emoji metadata to a JSON file
* **-J FILE** -- export emoji metadata to a JSON file in the format expected
  by Mutant Standard website code
* **-c** -- disable ANSI color codes; use if you see garbage in the terminal
  instead of pretty colorified output
* **-q WIDTHxHEIGHT** -- ensure source images have specified size
* **-t NUM** -- number of worker threads (default: **1**)
* **--force-desc** -- ensure all emoji have description (**desc** property)
* **-r RENDERER** -- selects rasteriser to use; the following rasterisers are
  supported: **inkscape**, **rendersvg**, **imagemagick**; **rendersvg** is
  recommended if speed is important; (default: **inkscape**)
* **-b NUM** -- maximum number of file arguments per exiftool call; larger
  numbers may accelerate metadata insertion but fail if the OS doesn't support
  sufficiently long argument lists; (default: **1000**)

Examples
--------

Export the entire emoji set as shortcode-named SVG files (default):
```
orxport.py
```

As above, but specifying manifest, input directory and output directory paths:

```
orxport.py -m <manifest file> -i <input directory> -o <output directory>
```

Export the entire emoji set as shortcode-named 64x64 and 256x256 PNG files,
with a separate directory for each format, using 4 threads:

```
orxport.py -F png-64,png-256 -f %f/%s -t 4
```

Export the entire emoji set as unicode-named SVG, 32x32 PNG, 128x128 PNG and
512x512 PNG files, using 8 threads, to a separate directory for each format,
ensuring that all input images are 32x32:

```
orxport.py -F svg,png-32,png-128,png-512 -f %f/%u -t 8 -q 32x32
```

Export only dark_elf emoji as a 256x256 PNG file:

```
orxport.py -F png-256 -e code=dark_elf
```

Export only emoji with v2 color modifier:

```
orxport.py -e color=v2
```

Export only paw emoji:

```
orxport.py -e morph=paw
```

Export all emoji as shortcode-named SVG files, preserving input directory
structure:

```
orxport.py -f %d/%s
```

Manifests
---------

Please refer to **MANIFEST.md** for documentation regarding manifests.


CHANGELOG.md
0.2.0
-----

* **rendersvg** rendering support
* **imagemagick** rendering support (thanks, @AndrewMontagne!)
* batch EXIF metadata insertion for efficiency (**-b** sets maximum batch size)
* **desc** property in emoji and colormaps
* **--force-desc** switch to ensure all emoji define **desc** property
* Automatically infer **root** if emoji isn't modifiable

0.1.0
-----

* First public release


export.py
import os
import pathlib
import queue
import re
import subprocess
import threading
import time

import log
import png
import svg
import util

class ExportThread:
    def __init__(self, queue, name, total, m, input_path, formats, path,
                 renderer):
        self.queue = queue
        self.name = name
        self.total = total
        self.m = m
        self.input_path = input_path
        self.formats = formats
        self.path = path
        self.renderer = renderer
        self.err = None
        self.kill_flag = False
        self.thread = threading.Thread(target=self.run)
        self.thread.start()

    def kill(self):
        self.kill_flag = True

    def join(self):
        self.thread.join()

    def msg(self, s, color=37, indent=0):
        log.out(s, color, indent, self.name)

    def export_svg(self, emoji_svg, path, license=None):
        if license:
            self.msg('* Writing license metadata...', indent=4)
            final_svg = svg.license(emoji_svg, license)
        else:
            final_svg = emoji_svg
        self.msg('* Exporting to file: ' + path, indent=4)
        try:
            out = open(path, 'w')
            out.write(final_svg)
            out.close()
        except Exception:
            raise Exception('Could not write to file: ' + path)

    def export_png(self, emoji_svg, size, path):
        self.msg('* Saving svg to temporary file...', indent=4)
        tmp_name = '.tmp' + self.name + '.svg'
        try:
            f = open(tmp_name, 'w')
            f.write(emoji_svg)
            f.close()
        except IOError:
            raise Exception('Could not write to temporary file: ' + tmp_name)
        self.msg(f'* Exporting at {size}px to {path}...', indent=4)
        if self.renderer == 'inkscape':
            cmd = ['inkscape', os.path.abspath(tmp_name),
                   '--export-png=' + os.path.abspath(path),
                   '-h', str(size), '-w', str(size)]
        elif self.renderer == 'rendersvg':
            cmd = ['rendersvg', '-w', str(size), '-h', str(size),
                    os.path.abspath(tmp_name), os.path.abspath(path)]
        elif self.renderer == 'imagemagick':
            cmd = ['convert', '-background', 'none', '-density', str(size / 32 * 128),
                   '-resize', str(size) + 'x' + str(size), os.path.abspath(tmp_name), os.path.abspath(path)]
        else:
            raise AssertionError
        try:
            r = subprocess.run(cmd, stdout=subprocess.DEVNULL).returncode
        except Exception as e:
            raise Exception('Rasteriser invocation failed: ' + str(e))
        if r:
            raise Exception('Rasteriser returned error code: ' + str(r))
        self.msg('* Deleting temporary file...', indent=4)
        os.remove(tmp_name)

    def export_emoji(self, emoji, emoji_svg, f, path, license):
        final_path = format_path(path, emoji, f)
        self.msg('* Export path is ' + final_path, indent=4)
        try:
            dirname = os.path.dirname(final_path)
            if dirname:
                os.makedirs(dirname, exist_ok=True)
        except IOError:
            raise Exception('Could not create directory: ' + dirname)
        if f == 'svg':
            self.export_svg(emoji_svg, final_path, license.get('svg'))
        elif f.startswith('png-'):
            try:
                size = int(f[4:])
            except ValueError:
                raise ValueError('Invalid format: ' + f)
            self.export_png(emoji_svg, size, final_path)
        else:
            raise ValueError('Invalid format: ' + f)

    def run(self):
        try:
            while not self.kill_flag:
                try:
                    i, emoji = self.queue.get_nowait()
                except queue.Empty:
                    break
                self.msg(f'[{i+1} / {self.total}] Exporting '
                         f'{emoji.get("code", "<UNNAMED>")}...', 32)
                try:
                    format_path(self.path, emoji, 'svg')
                except SkipException as ex:
                    if str(ex):
                        self.msg(f'Skipping: {ex}', 34, 4)
                    else:
                        self.msg('Skipping', 34, 4)
                    continue
                if 'src' not in emoji:
                    raise ValueError('Missing src attribute')
                srcpath = os.path.join(self.m.homedir, self.input_path,
                                       emoji['src'])
                self.msg('* Loading source file: ' + srcpath, indent=4)
                try:
                    emoji_svg = open(srcpath, 'r').read()
                except Exception:
                    raise ValueError('Could not load file: ' + srcpath)
                if 'color' in emoji:
                    self.msg('* Converting colormap...', indent=4)
                    cmap = self.m.colormaps[emoji['color']]
                    pfrom = self.m.palettes[cmap['src']]
                    pto = self.m.palettes[cmap['dst']]
                    emoji_svg = svg.ctrans(emoji_svg, pfrom, pto)
                for f in self.formats:
                    self.msg('-> ' + f, 35)
                    self.export_emoji(emoji, emoji_svg, f, self.path, self.m.license)
        except Exception as e:
            self.err = e

class SkipException(Exception):
    def __init__(self, s=''):
        self.s = s

    def __str__(self):
        return self.s

def format_resolve(code, emoji, f):
    if code[0] == '(':
        inside = code[1:-1]
        if inside not in emoji:
            raise ValueError('Missing property: ' + inside)
        return emoji[inside]
    if code == 'c':
        if 'color' not in emoji:
            raise ValueError('Cannot resolve %c - no colormap')
        return emoji['color']
    if code == 'd':
        if 'src' not in emoji:
            raise ValueError('Cannot resolve %d - no emoji source file defined')
        return str(pathlib.Path(emoji['src']).parent)
    if code == 'f':
        return f
    if code == 's':
        if 'code' not in emoji:
            raise ValueError('Cannot resolve %s - no shortcode')
        return emoji['code']
    if code == 'u':
        if 'unicode' not in emoji:
            raise ValueError('Cannot resolve %u - no unicode codepoint defined')
        if '!' in emoji['unicode']:
            raise SkipException('Cannot resolve %u (explicitly undefined)')
        return util.uni_to_hex_filename(emoji['unicode'])
    raise ValueError('Cannot resolve format code: ' + code)

def format_path(path, emoji, f):
    res = path
    if f == 'svg':
        res = res + '.svg'
    elif f.startswith('png-'):
        res = res + '.png'
    else:
        raise ValueError('Invalid export format: ' + f)
    for match, fcode in set(re.findall(r'(%(\(.*\)|.))', res)):
        repl = format_resolve(fcode, emoji, f)
        res = res.replace(match, repl)
    return res

def export(m, filtered_emoji, input_path, formats, path, src_size,
           num_threads, renderer, max_batch):
    # 1st pass
    log.out('Performing sanity check...', 36)
    for i, e in enumerate(filtered_emoji):
        log.out(f'[{i+1} / {len(filtered_emoji)}] Checking '
                f'{e.get("code", "<UNNAMED>")}...', 32)
        try:
            format_path(path, e, 'svg')
        except SkipException as ex:
            if str(ex):
                log.out(f'Skipping: {ex})', 34, 4)
            else:
                log.out('Skipping', 34)
            continue
        if 'src' not in e:
            raise ValueError('Missing src attribute')
        srcpath = os.path.join(m.homedir, input_path, e['src'])
        try:
            emoji_svg = open(srcpath, 'r').read()
        except Exception:
            raise ValueError('Could not load file: ' + srcpath)
        if src_size is not None:
            imgsize = svg.size(emoji_svg)
            if imgsize != src_size:
                raise ValueError('Source image size is {}, expected {}'.format(
                    str(imgsize[0]) + 'x' + str(imgsize[1]),
                    str(src_size[0]) + 'x' + str(src_size[1])))
    # 2nd pass
    log.out('Exporting emoji...', 36)
    emoji_queue = queue.Queue()
    for entry in enumerate(filtered_emoji):
        emoji_queue.put(entry)
    log.show_threads = num_threads > 1
    threads = []
    for i in range(num_threads):
        log.out(f'Init thread {i}...', 35)
        threads.append(ExportThread(emoji_queue, str(i), len(filtered_emoji),
                                    m, input_path, formats, path, renderer))
    while True:
        done = emoji_queue.empty()
        for t in threads:
            if t.err is not None:
                for u in threads:
                    u.kill()
                    u.join()
                raise ValueError(f'Thread {t.name} failed: {t.err}')
        if done:
            break
        time.sleep(0.01)
    log.out(f'Waiting for all threads to finish...', 35)
    for t in threads:
        t.join()
    # png license pass
    if 'png' in m.license:
        png_files = []
        for e in filtered_emoji:
            for f in formats:
                if f.startswith('png-'):
                    try:
                        png_files.append(format_path(path, e, f))
                    except SkipException:
                        continue
        log.out(f'Adding license metadata to {len(png_files)} png files...', 36)
        png.license(png_files, m.license.get('png'), max_batch)
